### PR TITLE
Convert to python3

### PR DESCRIPTION
**Pull request by moon-chilled**
There are still some UnicodeDecodeErrors I'm confused about, but overall, it seems to work.

### CODE REVIEW

File: asciidoc.py [hunk 1]
```diff
@@ -1296,9 +1264,13 @@ class Lex:
...
     def __init__(self):
-        raise AssertionError,'no class instances allowed'
+        raise AssertionError('no class instances allowed')
+
+    def __iter__(self):
+        return self
+
     @staticmethod
-    def next():
+    def __next__():
```
Comment (elextr) [inline asciidoc.py]: I think this is better named `next_element()` not `__next__()` since it is not the iterator next function.

**Comment by elextr** [comment]
Generally looks ok, nice job.

I'm not sure about your encod/decode removals, are you sure you havn't removed the ability for input files and output files to be in non-UTF-8 encodings?

Now a tricky question, since Python2 regexes defaulted to ASCII match and Python 3 regexes default to Unicode match have you checked every regex in asciidoc and the config files still works as intended, those with the `u` flag set will be fine but not all of them had the `u` flag set.  The question is did those without the `u` flag intend ASCII match and so they need the `a` flag set now, or do they not care if its ASCII or Unicode and they are fine.

**Comment by Elronnd** [comment]
> I'm not sure about your encod/decode removals, are you sure you havn't removed the ability for input files and output files to be in non-UTF-8 encodings?

I'm pretty sure that non-utf8 won't work anymore because python's `str` is utf-8 only.  I can go back and make it binary read/manual decode if necessary.

> Now a tricky question, since Python2 regexes defaulted to ASCII match and Python 3 regexes default to Unicode match have you checked every regex in asciidoc and the config files still works as intended

Uh...no, I actually didn't know about that change.  I can take a look but my intuition says that everything will *probably* continue to work.

**Comment by elextr** [comment]
> > I'm not sure about your encod/decode removals, are you sure you havn't removed the ability for input files and output files to be in non-UTF-8 encodings?

> I'm pretty sure that non-utf8 won't work anymore because python's str is utf-8 only. I can go back and make it binary read/manual decode if necessary.

I am working on another implementation of Asciidoc and what we do is convert all inputs to utf-8 and use that internally, which would be the same as you converting to Python unicode strings on input.

In my unscientific assessment, its input thats most important and thats easiest as I outlined above.  Outputs are HTML/XML and UTF-8 should be fine IMHO, but others may disagree.

Alternatively you can just make an issue to add output encodings later, stating its currently UTF-8 only, but it would be nice to accept non-UTF-8 inputs to allow for existing documents.

> > Now a tricky question, since Python2 regexes defaulted to ASCII match and Python 3 regexes default to Unicode match have you checked every regex in asciidoc and the config files still works as intended

> Uh...no, I actually didn't know about that change. I can take a look but my intuition says that everything will probably continue to work.

IIUC its mostly the backslash ones that matter eg `\w` `\W` ASCII vs Unicode word chars.  In most cases its actually an improvement that they work with all Unicode word characters and so on, but some constructs need to be ASCII still, such as IDs.

The easiest way to do pretty comprehensive regression testing is to run the Python2 version of asciidoc to generate the reference output for comparison with the Python 3 version, see [here](http://asciidoc.org/testasciidoc.html).

**Comment by Elronnd** [comment]
TOTAL PASSED:  162
    TOTAL FAILED:  40

:cry:

It's getting pretty late here, but I'll take a look at what might be causing those tomorrow.

**Comment by elextr** [comment]
No problem, once it seems to be working reasonably will merge this PR and then normal bugfix service can be resumed as people try it and test it.

**Comment by elextr** [comment]
@Elronnd hows it going getting the regression tests to pass?

**Comment by odysseusPolytropos** [comment]
Hey guys,

what is the situation of asciidoc with python3? As far as I can see from the comments/commits @Elronnd s fork for this pull request ist the last (and most far) step for using asciidoc with pyhton3. Is that correct? Are there any plans to merge this PR in near future or is here some help with the regressions needed? If so, I would take a look, because I would prefer using asciidoc with python3 instead of downgrading to python2. And working on this branch seems still reasonable, since there has not been much development on the original https://github.com/asciidoc/asciidoc repository, right?

Cheers,
odysseus

**Comment by elextr** [comment]
@odysseusPolytropos no its the first step, not the last step.

First question, why not use the Asciidoctor implementation? Its actively developed and supported.

My understanding was the OP was going to look at the tests that failed, but nothing further has been heard, even after the ping above.  

The Python 2 version is stable so I don't want a repository with a Python 3 version with many possible subtle bugs (40 failing tests) that is unsupported but users may make the same mistake as you and try to use for real work.  That will just give the whole Asciidoc infrastructure a bad name.

If you want to take the pull request and make it pass the tests that would be welcome, but you would be lumbered with supporting it, the active Asciidoc implementation is now Asciidoctor and most contributors have moved to that.

**Comment by fenrus75** [comment]
asciidoctor looks great. Until I saw the ruby word. Now I need yet another whole language ecosystem installed/packaged/etc.

**Comment by fenrus75** [comment]
yup I just packaged it

I had to do

$ cat build.patch
--- asciidoc-py3-cd6762bd16b20e5cf19ceaffb8452df912bb5f6e/Makefile.in~
2018-06-12 21:57:09.000000000 +0000
+++ asciidoc-py3-cd6762bd16b20e5cf19ceaffb8452df912bb5f6e/Makefile.in
 2018-06-16 17:18:09.807598205 +0000
@@ -107,7 +107,7 @@
        $(INSTALL_DATA) $($@) $(DESTDIR)/$($<)/

 $(manp): %.1 : %.1.txt
-       python2 a2x.py -f manpage $<
+       python3 a2x.py -f manpage $<

 docs:
        $(INSTALL) -d $(DESTDIR)/$(docdir)


to build it but that's little details ;)

**Comment by elextr** [comment]
Thanks for passing that on. IIUC its been packaged by Fedora, wonder why they didn't find that, or why they didn't report it?

**Pull request merged**